### PR TITLE
Optimization of CheckStatusCodeHandler in NancyEngine

### PR DESCRIPTION
**Pull request by thecodejunkie**
Minor tweak while chasing bigger dragons. The `CheckStatusCodeHandler` runs on every request to see if it should be taking any special action based on the returned status code. This is part of a 100k requests sample

- Removed LINQ magic

This loops over the `IStatusCodeHandler` once and mimics of a `FirstOrDefault` by only assigning values if nothing has previously been written. If both the `default` and `custom` handlers has been assigned, then it will exist the loop early

Before
image

After
image

### CODE REVIEW

File: src/Nancy/NancyEngine.cs [hunk 1]
```diff
@@ -217,17 +217,42 @@ private void CheckStatusCodeHandler(NancyContext context)
...
+                    {
+                        break;
+                    }
+                }
+            }
 
-            var customHandler = handlers
-                .FirstOrDefault(x => !(x is DefaultStatusCodeHandler));
+            //var handlers = this.statusCodeHandlers
+            //    .Where(x => x.HandlesStatusCode(context.Response.StatusCode, context))
```
Comment (khellang) [inline src/Nancy/NancyEngine.cs]: :scream_cat:

File: src/Nancy/NancyEngine.cs [hunk 2]
```diff
@@ -217,17 +217,42 @@ private void CheckStatusCodeHandler(NancyContext context)
...
 
-            var defaultHandler = handlers
-                .FirstOrDefault(x => x is DefaultStatusCodeHandler);
+            foreach (var statusCodeHandler in this.statusCodeHandlers)
+            {
+                if (statusCodeHandler.HandlesStatusCode(context.Response.StatusCode, context))
+                {
+                    if (defaultHandler == null && (statusCodeHandler is DefaultStatusCodeHandler))
+                    {
+                        defaultHandler = statusCodeHandler;
```
Comment (khellang) [inline src/Nancy/NancyEngine.cs]: couldn't you just `continue` here?

File: src/Nancy/NancyEngine.cs [hunk 3]
```diff
@@ -217,17 +217,42 @@ private void CheckStatusCodeHandler(NancyContext context)
...
+                    }
+
+                    if (customHandler == null && !(statusCodeHandler is DefaultStatusCodeHandler))
+                    {
+                        customHandler = statusCodeHandler;
```
Comment (khellang) [inline src/Nancy/NancyEngine.cs]: same here. it's not possible that a handler can be both custom and default, right?

File: src/Nancy/NancyEngine.cs (hunk 1 above)
Comment (thecodejunkie) [inline src/Nancy/NancyEngine.cs]: Whoops :astonished:

File: src/Nancy/NancyEngine.cs [hunk 4]
```diff
@@ -83,7 +83,7 @@ public class NancyEngine : INancyEngine
 
             this.dispatcher = dispatcher;
             this.contextFactory = contextFactory;
-            this.statusCodeHandlers = statusCodeHandlers;
+            this.statusCodeHandlers = statusCodeHandlers.ToArray();
```
Comment (phillip-haydon) [inline src/Nancy/NancyEngine.cs]: Calling `ToArray()` will copy the list passed in, can't we just make the argument array to begin with and avoid the copy?

File: src/Nancy/NancyEngine.cs [hunk 5]
```diff
@@ -217,17 +217,34 @@ private void CheckStatusCodeHandler(NancyContext context)
...
+            IStatusCodeHandler customHandler = null;
 
-            var defaultHandler = handlers
-                .FirstOrDefault(x => x is DefaultStatusCodeHandler);
-
-            var customHandler = handlers
-                .FirstOrDefault(x => !(x is DefaultStatusCodeHandler));
+            foreach (var statusCodeHandler in this.statusCodeHandlers)
+            {
+                if (statusCodeHandler.HandlesStatusCode(context.Response.StatusCode, context))
```
Comment (phillip-haydon) [inline src/Nancy/NancyEngine.cs]: Can we invert the if so we don't have such nested code. 

```
if (!...) 
{
continue;
}

// stuffs thats inside the if
```

**Pull request merged**